experimental/sdata/frame.cue

```diff
@@ -0,0 +1,54 @@
+package sdata
+
+import "list"
+
+#FieldType: "time" | "number" | "string"
+#FieldTypeInfoFrame: "time.Time" | "float64" | "string"  | "int64"
+#FieldTypeInfo: {
+	frame: #FieldTypeInfoFrame
+}
+#FieldSchema: {
+	name: string
+	type: #FieldType
+	typeInfo: #FieldTypeInfo
+	labels?: {
+		[string]: string
+	}
+}
+#FieldValue: string | number
+#FieldValues: [#FieldValue, ...]
+#FieldTypeInfoMap: {
+	"time.Time": int
+	float64: number 
+	int64: int
+	"string": string
+}
+
+#FrameType: "timeseries-many" | "timeseries-wide" | "timeseries-long"
+#FrameMeta: {
+	type: #FrameType
+}
+#FrameSchema: {
+	name?: string
+	fields: [#FieldSchema, ...]
+	meta: #FrameMeta
+}
+#FrameData: {
+	values: [#FieldValues, ...]
+	#expectedLength: [ 
+		for v in values { 
+			v & list.MinItems(len(values[0])) & list.MaxItems(len(values[0])) 
+		}
+	]
+}
+#Frame: {
+	schema: #FrameSchema
+	data: #FrameData
+	#matchingTypes: [
+		for i, fv in data.values { 
```

Comment:
is it possible to generate go code for validations like this?

Reply:
makes sense, thanks for following up. i'll have to mess with CUE more at some point and mess with generation

Reply:
meant to write that comment ages ago - the CLA assistant reminded me of this issue :)

Happy to help when you get around to it. The upstream CUE libraries won't be helpful for Go codegen, especially at the level of sophistication you're interested in here. With what i've grokked here so far, i would guess it'll be a hybrid of handwritten types, plausibly generic, plus some codegen.

Having done a few rounds of such things myself now, i don't think it should be horribly bad. But i have lots of scars now, and my guess is that you'd have a terrible time if you try to [leeroy](https://knowyourmeme.com/memes/leeroy-jenkins) it on your own ;)